### PR TITLE
Add new modules to support uploading Fence and Rally with Mission Item Protocol

### PR DESCRIPTION
**Pull request by peterbarker**
.... also changes to mavproxy_map to allow manipulation of same.

### CODE REVIEW

**Comment by tridge** [comment]
@peterbarker ping?

**Comment by tridge** [comment]
@peterbarker it seems to be missing some changes in mavwp: ```
Traceback (most recent call last):
  File "/home/tridge/.local/bin/mavproxy.py", line 4, in <module>
    __import__('pkg_resources').run_script('MAVProxy==1.8.65', 'mavproxy.py')
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 656, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 1460, in run_script
    exec(script_code, namespace, namespace)
  File "/home/tridge/.local/lib/python3.10/site-packages/MAVProxy-1.8.65-py3.10.egg/EGG-INFO/scripts/mavproxy.py", line 1448, in <module>
  File "/home/tridge/.local/lib/python3.10/site-packages/MAVProxy-1.8.65-py3.10.egg/EGG-INFO/scripts/mavproxy.py", line 530, in load_module
  File "/home/tridge/.local/lib/python3.10/site-packages/MAVProxy-1.8.65-py3.10.egg/MAVProxy/modules/mavproxy_rally.py", line 143, in init
  File "/home/tridge/.local/lib/python3.10/site-packages/MAVProxy-1.8.65-py3.10.egg/MAVProxy/modules/mavproxy_rally.py", line 21, in __init__
AttributeError: module 'pymavlink.mavwp' has no attribute 'MissionItemProtocol_Rally'
```
maybe we should just not load rally module if this fails?

File: MAVProxy/modules/lib/mission_item_protocol.py [hunk 1]
```diff
@@ -0,0 +1,1077 @@
...
+        # preserve srcSystem as that is used for naming waypoint file
+        wp2._header.srcSystem = wp.get_srcSystem()
+        wp2._header.srcComponent = wp.get_srcComponent()
+        return wp2
+
+    def process_waypoint_request(self, m, master):
+        '''process a waypoint request from the master'''
+        if m.mission_type != self.mav_mission_type():
+            return
+        print("Processing %s request: (%s)" % (self.itemtype(), str(m)))
```
Comment (tridge) [inline MAVProxy/modules/lib/mission_item_protocol.py:383]: this debug is very annoying

File: MAVProxy/modules/lib/mission_item_protocol.py [hunk 2]
```diff
@@ -0,0 +1,1077 @@
...
+        if getattr(self.wploader, 'expected_count', None) is not None:
+            self.wploader.expected_count = 0
+        self.loading_waypoint_lasttime = time.time()
+
+    def cmd_list(self, args):
+        self.wp_op = "list"
+        self.request_list_send()
+
+    def cmd_load(self, args):
+        print("args: %s" % str(args))
```
Comment (tridge) [inline MAVProxy/modules/lib/mission_item_protocol.py:799]: remove debug

File: MAVProxy/modules/lib/mission_item_protocol.py [hunk 3]
```diff
@@ -0,0 +1,1077 @@
...
+
+    def pretty_parameter_value(self, value):
+        '''pretty parameter value'''
+        return value
+
+    def savecsv(self, filename):
+        '''save waypoints to a file in human-readable CSV file'''
+        f = open(filename, mode='w')
+        headers = ["Seq", "Frame", "Cmd", "P1", "P2", "P3", "P4", "X", "Y", "Z"]
+        print(self.csv_line(headers))
```
Comment (tridge) [inline MAVProxy/modules/lib/mission_item_protocol.py:915]: needed?

File: MAVProxy/modules/lib/mission_item_protocol.py [hunk 4]
```diff
@@ -0,0 +1,1077 @@
...
+                self.pretty_enum_value('MAV_CMD', w.command),
+                self.pretty_parameter_value(w.param1),
+                self.pretty_parameter_value(w.param2),
+                self.pretty_parameter_value(w.param3),
+                self.pretty_parameter_value(w.param4),
+                self.pretty_parameter_value(w.x),
+                self.pretty_parameter_value(w.y),
+                self.pretty_parameter_value(w.z),
+            ]
+            print(self.csv_line(out_list))
```
Comment (tridge) [inline MAVProxy/modules/lib/mission_item_protocol.py:933]: more debug

File: MAVProxy/modules/lib/mission_item_protocol.py [hunk 5]
```diff
@@ -0,0 +1,1077 @@
...
+            self.target_component,
+            mission_type=self.mav_mission_type())
+
+    def wp_ftp_download(self, args):
+        '''Download wpts from vehicle with ftp'''
+        ftp = self.mpstate.module('ftp')
+        if ftp is None:
+            print("Need ftp module")
+            return
+        print("Fetching mission with ftp")
```
Comment (tridge) [inline MAVProxy/modules/lib/mission_item_protocol.py:955]: not needed

**Review by tridge — CHANGES_REQUESTED** [review]
the FEN icon in console doesn't seem to be updating correctly? Shows as grey with a fence enabled and in breach

**Review by peterbarker — COMMENTED** [review]
Fixed

**Comment by peterbarker** [comment]
> the FEN icon in console doesn't seem to be updating correctly? Shows as grey with a fence enabled and in breach

This turned out to be because the action was report-only.

**Review by tridge — APPROVED** [review]

**Pull request merged**